carma_guidance_plugins/include/carma_guidance_plugins/plugin_base_node.hpp

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2022 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#pragma once
+
+#include <gtest/gtest_prod.h>
+#include <rclcpp/rclcpp.hpp>
+#include <carma_planning_msgs/msg/plugin.hpp>
+#include <carma_wm_ros2/WMListener.hpp>
+#include <carma_wm_ros2/WorldModel.hpp>
+
+#include <carma_ros2_utils/carma_lifecycle_node.hpp>
+
+namespace carma_guidance_plugins
+{
+
+  /**
+   * \brief PluginBaseNode provides default functionality for all carma guidance plugins.
+   *        This includes basic state machine management (largely delegated to ROS2 lifecycle behavior), required interfaces, and plugin discovery
+   * 
+   * Extending classes must implement the on_configure_plugin method to load parameters, and my override the other state transitions methods on_<state>_plugin if desired.
+   * Additionally, extending classes must implement the methods such as get_name() which are used to populate the plugin discovery message.
+   * 
+   */
+  class PluginBaseNode : public carma_ros2_utils::CarmaLifecycleNode
+  {
+
+  private:
+    // Publishers
+    //! Publisher for plugin discovery. This publisher will begin publishing 
+    //immediately on node construction regardless of lifecycle state. This is meant to allow for fast plugin activation on startup.
+    rclcpp::Publisher<carma_planning_msgs::msg::Plugin>::SharedPtr plugin_discovery_pub_;
+
+    // Timers
+    //! Timer to trigger publication of the plugin discovery message at a fixed frequency 
+    rclcpp::TimerBase::SharedPtr discovert_timer_;
+
+    // WorldModel listener
+    carma_wm::WMListener wm_listener_;
```

Comment:
Is there overhead in having this in the base? What's the likelihood that a plugin doesn't need the WM? Doesn't seem super likely outside of really simple cases right?

Reply:
There is definitely some overhead. I went back and forth on it, particularly for the control plugins which currently don't use it. 
I would personally be ok either way. It's only 2-3 extra lines for someone to do it manually. 
Any preference?

Reply:
Is it possible to somehow make it so they're optionally defined in the class? Seems like the sort of thing templates should be able to support, but I'm not sure how exactly that'd work. Could definitely do it with an `#ifdef __USE_WM__ ...` macro block but that's kinda clunky too. I think if that's too much of a lift, let's just this down from the base plugin class to the strategic and tactical bases. We can always refactor it back to the base later if we find ourselves wanting the WM in control plugins.

Reply:
I'll move it to the strategic and tactical bases for now. Having a macro like that would require the user to modify the cmake file to set it which is not something most of the team does on the regular and would be confusing to others looking at example plugins.

Reply:
Scratch that,
Per discussion on slack we will lazily initialize the world model listener as a way of only creating subscriptions if the plugin is using that variable.

Reply:
done